server/middlewares/rateLimiter.ts

```diff
@@ -0,0 +1,63 @@
+import { Context, Next } from "koa";
+import { defaults } from "lodash";
+import RateLimiter from "@server/RateLimiter";
+import env from "@server/env";
+import { RateLimitExceededError } from "@server/errors";
+import Logger from "@server/logging/Logger";
+import Redis from "@server/redis";
+import { RateLimiterConfig } from "@server/types";
+
+export function rateLimiter() {
+  return async function rateLimiterMiddleware(ctx: Context, next: Next) {
+    if (!env.RATE_LIMITER_ENABLED) {
+      return next();
+    }
+
+    const limiter = RateLimiter.getRateLimiter(ctx.path);
+
+    try {
+      await limiter.consume(`${ctx.path}:${ctx.ip}`);
+    } catch (rateLimiterRes) {
+      ctx.set("Retry-After", `${rateLimiterRes.msBeforeNext / 1000}`);
+      ctx.set("RateLimit-Limit", `${limiter.points}`);
+      ctx.set("RateLimit-Remaining", `${rateLimiterRes.remainingPoints}`);
+      ctx.set(
+        "RateLimit-Reset",
+        `${new Date(Date.now() + rateLimiterRes.msBeforeNext)}`
+      );
+
+      throw RateLimitExceededError();
+    }
+
+    return next();
+  };
+}
+
+export function registerRateLimiter(config: RateLimiterConfig) {
+  return async function registerRateLimiterMiddleware(
+    ctx: Context,
+    next: Next
+  ) {
+    if (!env.RATE_LIMITER_ENABLED) {
+      return next();
+    }
+
+    if (!RateLimiter.hasRateLimiter(ctx.path)) {
+      // kill the default rate limiter for this path first
```

Comment:
I'm sure it is, but it's not clear to me why this is required – I'm _guessing_ you were finding it hard to debug turning route-specific limits on/off?

Reply:
My bad! Should've explained the _why_.

> I'm guessing you were finding it hard to debug turning route-specific limits on/off?

In a way, yes. The new route-specific configuration doesn't apply unless the old keys in Redis are flushed. One way to solve this is to flush the old keys. Another solution is to specify a different `keyPrefix` from that of the default rate limiter. In the latter case, however, the old keys are not flushed forcibly, albeit upon `ttl` expiry.

Reply:
>True! I've already observed it a couple of times in dev. We have the option to either do fire and forget or not do it at all. There are apparent individual trade offs in each of the two approaches. In the former, we tend to trade off performance (consider having large number of keys) for determinism(new rate limiter configurations kick in immediately upon server restart). And, in the latter, it's vice versa.

>That being said, I tend to favor the latter one since determinism in the former is still not a guarantee because of what you pointed out. If you agree with me, I can get rid of it. We can put it back later once we have a pattern in place to handle shutdown gracefully.

Quoting from [here](https://github.com/outline/outline/pull/3857#discussion_r938453895). If we're fine sacrificing determinism here as well, I can get rid of this too. Let me know your thoughts.